distributed_training/validator/reward.py

```diff
@@ -416,74 +413,86 @@ def benchmark_untested_uids(self):
+                    if status == "SUCCESS":
+                        self.uid_tracker[uid]["all_reduce_successes"] += 1
                 else:
-                    self.uid_tracker[int(uid)]["all_reduce_score"] = 0
+                    bt.logging.warning(f"UID {uid} from allreduce_status_dict not found in uid_tracker.")
+            self.allreduce_status_dict = {}
+        else:
+            bt.logging.info("No AllReduce status to update or allreduce_status_dict not found.")
 
     except Exception as e:
-        bt.logging.info(f"Error {e} updating all_reduce scores")
+        bt.logging.error(f"Error updating all_reduce stats: {e}")
 
 
 def update_total_scores(self):
-    update_all_reduce_scores(self)
+    # Update AllReduce stats from the latest round
+    update_all_reduce_stats(self)
+
+    # Recalculate individual scores
+    for uid_key in self.uid_tracker:
+        uid_data = self.uid_tracker[uid_key]
+        if uid_data["all_reduce_counts"] > 0:
+            uid_data["all_reduce_score"] = (
+                uid_data["all_reduce_successes"]
+                / uid_data["all_reduce_counts"]
+            )
+        else:
+            uid_data["all_reduce_score"] = 0.0
+            
     # Sort uid tracker
     self.uid_tracker = dict(sorted(self.uid_tracker.items()))
 
     # Normalise each type of reward
-    train_score_normalised = np.linalg.norm(
-        [self.uid_tracker[uid]["train_score"] for uid in self.uid_tracker],
-        ord=1,
-        axis=0,
-        keepdims=True,
-    ).item()
-    all_reduce_score_normalised = np.linalg.norm(
-        [self.uid_tracker[uid]["all_reduce_score"] for uid in self.uid_tracker],
-        ord=1,
-        axis=0,
-        keepdims=True,
-    ).item()
-    repo_valid_score_normalised = np.linalg.norm(
-        [self.uid_tracker[uid]["repo_valid_score"] for uid in self.uid_tracker],
-        ord=1,
-        axis=0,
-        keepdims=True,
-    ).item()
-
-    # Check if any norms are zero or contains NaN values
-    if np.any(train_score_normalised == 0) or np.isnan(train_score_normalised).any():
-        train_score_normalised = np.ones_like(train_score_normalised).item()
-
-    if (
-        np.any(all_reduce_score_normalised == 0)
-        or np.isnan(all_reduce_score_normalised).any()
-    ):
-        all_reduce_score_normalised = np.ones_like(all_reduce_score_normalised).item()
-
-    if (
-        np.any(repo_valid_score_normalised == 0)
-        or np.isnan(repo_valid_score_normalised).any()
-    ):
-        repo_valid_score_normalised = np.ones_like(repo_valid_score_normalised).item()
+    all_train_scores = [
+        self.uid_tracker[uid].get("train_score", 0.0) for uid in self.uid_tracker
+    ]
+    all_all_reduce_scores = [
+        self.uid_tracker[uid].get("all_reduce_score", 0.0) for uid in self.uid_tracker
+    ]
+    all_repo_valid_scores = [
+        self.uid_tracker[uid].get("repo_valid_score", 0.0) for uid in self.uid_tracker
+    ]
+
+    train_score_norm = (
```

Comment:
@jundi69 FYI this change had a bug where the new `all_reduce_score_norm` variable was a list vs before it was a number since we ran .item() on the output of `np.linalg.norm()`. This then caused `self.uid_tracker[uid]["total_score"]` to be a list which then caused the validator to crash when running: https://github.com/KMFODA/DistributedTraining/blob/a3015d35694a853252cc245c10c1e81c6cc6c3d2/distributed_training/base/validator.py#L429.

This was fixed before Friday's merge but it would be good to test for these types of errors before pushing commits like these. Just running a non_master validator on this commit for 1 or 2 steps would have highlighted this error.